File: 279. Perfect Squares.js
/* https://leetcode.com/problems/perfect-squares/ */

/**
 * @param {number} n
 * @return {number}
 */

var numSquares = function (n) {
	let squares = [];
	let i,
		j = parseInt(Math.sqrt(n));

	while (j) {
		squares.push(j * j);
		j--;
	}

	let dict = {};

	let compute = val => {
		if (val <= 0) {
			return 0;
		}

		if (val in dict) {
			return dict[val];
		}

		j = Infinity;
		for (let i = 0; i < squares.length; i++) {
			if (0 <= val - squares[i]) {
				j = Math.min(j, compute(val - squares[i]));
			}
		}

		dict[val] = j === Infinity ? 0 : j + 1;
		return dict[val];
	};

	compute(n);
	return dict[n];
};

console.log(numSquares(89061));
console.log(numSquares(13));

File: 279. Perfect Squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        import math

        dict = {}
        squares = []
        j = int(math.sqrt(n))

        while 0 < j:
            squares.append(j * j)
            j -= 1

        def compute(val):
            if val <= 0:
                return 0
            if val in dict:
                return dict[val]

            j = val

            for v in squares:
                if 0 <= val - v:
                    j = min(j, compute(val - v))

            dict[val] = j + 1 if j != val else val
            return dict[val]

        compute(n)
        return dict[n]


n = Solution()
print(n.numSquares(8328))
